lib/smart_proxy_container_gateway/container_gateway_main.rb

```diff
@@ -57,140 +59,124 @@ def tags(repository, params = {})
+        query = params[:q]
+        query = nil if query == ''
 
-          if params[:q].nil? || params[:q] == "" || repo_name.include?(params[:q])
-            repo_count += 1
-            repositories << { name: repo_name }
-          end
-        end
-        repositories
+        user = params[:user].nil? ? nil : database[:useres].find(name: params[:user])
+
+        # TODO: relies on catalog returning a query
+        repositories = query ? catalog(user).filter(:name.like("%#{query}%")) : catalog(user)
+        repositories.limit(limit)
       end
 
       def catalog(user = nil)
         if user.nil?
           unauthenticated_repos
         else
-          (unauthenticated_repos + user.repositories_dataset.map(:name)).sort
+          # TODO implement as a single query
+          unauthenticated_repos | user.repositories_dataset.select(:name)
         end
       end
 
       def unauthenticated_repos
-        Repository.where(auth_required: false).order(:name).map(:name)
+        database[:repository].where(auth_required: false).order(:name).select(:name)
       end
 
       # Replaces the entire list of repositories
       def update_repository_list(repo_list)
-        RepositoryUser.dataset.delete
-        Repository.dataset.delete
-        repo_list.each do |repo|
-          Repository.find_or_create(name: repo['repository'],
-                                    auth_required: repo['auth_required'].to_s.downcase == "true")
-        end
+        # TODO: make repositories_users cascade on delete repository
+        database[:repositories_users].delete
+
+        repository = database[:repository]
+        repository.delete
+
+        repository.import(
+          [:name, :auth_required],
+          repo_list.map { |repo| [repo['repository'], repo['auth_required'].to_s.downcase == "true"] },
+        )
       end
 
       # Replaces the entire user-repo mapping for all logged-in users
       def update_user_repo_mapping(user_repo_maps)
         # Get hash map of all users and their repositories
         # Ex: {"users"=> [{"admin"=>[{"repository"=>"repo", "auth_required"=>"true"}]}]}
         # Go through list of repositories and add them to the DB
-        RepositoryUser.dataset.delete
-        user_repo_maps['users'].each do |user_repo_map|
-          user_repo_map.each do |user, repos|
-            next if repos.nil?
-
-            repos.each do |repo|
-              found_repo = Repository.find(name: repo['repository'],
-                                           auth_required: repo['auth_required'].to_s.downcase == "true")
-              if found_repo.nil?
-                logger.warn("#{repo['repository']} does not exist in this smart proxy's environments")
-              elsif found_repo.auth_required
-                found_repo.add_user(User.find(name: user))
-              end
-            end
+        repositories = database[:repositories]
+
+        entries = user_repo_maps['users'].flat_map do |user_repo_map|
```

Comment:
I don't know if import can handle this flat map - you may need to run it per user